### PR TITLE
usm: npm: Unify protocol classification

### PR DESCRIPTION
**Pull request by guyarb**
### What does this PR do?

The new method will allow us to share classification among the different products.
The benefits are:
  1. Avoid duplicated processing of packets by sharing a common classification map.
  2. Allow NPM to have the underlying protocol when TLS is introduced (if supported and enabled by USM).

In the process, I had to make the following changes:
  1. Move `clean_protocol_classification` from `cleanup_conn`, to be called in `kretprobe__tcp_close`.
     The reason for that, is synchronization with the socket filters, as `kprobe__tcp_close` is
     called before the socket filter, we cannot delete the connection protocol information from
     the map, otherwise, we'll try to classify the connection termination, since there is no
     cached protocol. The impact is mainly on USM, as in that scenario, we're terminating pending
     http requests, and we're loosing the last request in a connection.
  2. Split `kretprobe__tcp_close` into `kretprobe__tcp_close_clean_protocols` and `kretprobe__tcp_close_flush`. The first hook is responsible for calling `cleanup_conn`, the latter contains the original implementation of `kretprobe__tcp_close`. I had to split the logic into 2 functions, as `flush_conn_close_if_full` uses almost all 512 bytes allowed per stack in eBPF. If protocol classification is enabled and supported, then we set `kretprobe__tcp_close_clean_protocols` as the return hook of `tcp_close` and we use `tail call` to trigger `kretprobe__tcp_close_flush`, and if it is not enabled or supported, then `kretprobe__tcp_close_flush` is the return hook.
  3. If protocol classification is supported and enabled, then in `kprobe__tcp_close` we're caching in a dedicate map the `conn_tuple_t`, and in `kretprobe__tcp_close`
     we're fetching it. That's being done as the exit functions do not have access to input arguments.

The changes were not applied to `fentry` hooks.
 

### Motivation
1. Reduce processing time of classification.
2. Allow NPM to have the underlying protocol if USM is enabled.

### Additional Notes

MAX cpu - small bump (2%)
image

AVG cpu - no diff
image

AVG RSS - small (1%) improvement
image

MAX RSS - small (1%) improvement
image

### Describe how to test/QA your changes

1. Enable USM 
2. Run an HTTPs server
3. Run an HTTPs client
5. Grab connections, make sure the protocol stack is TLS and then HTTP

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Review by hmahmood — APPROVED** [review]

**Comment by guyarb** [comment]
CI is failing due to https://github.com/DataDog/datadog-agent/pull/16714

File: pkg/network/tracer/connection/kprobe/config.go [hunk 1]
```diff
@@ -51,7 +52,7 @@ func enabledProbes(c *config.Config, runtimeTracer bool) (map[probes.ProbeFuncNa
 		enableProbe(enabled, probes.TCPReadSock)
 		enableProbe(enabled, probes.TCPReadSockReturn)
 		enableProbe(enabled, probes.TCPClose)
-		enableProbe(enabled, probes.TCPCloseReturn)
+		enableProbe(enabled, probes.TCPCloseFlushReturn)
```
[the remark this answers is not in the document]
Comment (guyarb) [reply pkg/network/tracer/connection/kprobe/config.go:55]: yes
unfortunately the current code does not distinguish between "not excluded" and activated.
so we're enabling that, and if `kretprobe__tcp_close_clean_protocols` is enabled as well, then we have the classification tail calls, and there is a code to "deactivate" enabled hooks.

hope I was clear enough

File: pkg/network/tracer/connection/kprobe/config.go [hunk 2]
```diff
@@ -40,6 +40,7 @@ func enabledProbes(c *config.Config, runtimeTracer bool) (map[probes.ProbeFuncNa
 			enableProbe(enabled, probes.ProtocolClassifierQueuesSocketFilter)
 			enableProbe(enabled, probes.ProtocolClassifierDBsSocketFilter)
 			enableProbe(enabled, probes.NetDevQueue)
+			enableProbe(enabled, probes.TCPCloseCleanProtocolsReturn)
```
[the remark this answers is not in the document]
Comment (guyarb) [reply pkg/network/tracer/connection/kprobe/config.go:43]: we can, but it introduces redundant operations that are not relevant
I'm under the impression that if we can "spare redundant functionality, we should"

**Review by hmahmood — APPROVED** [review]

File: pkg/network/protocols/http/ebpf_main.go [hunk 3]
```diff
@@ -192,12 +193,13 @@ func newEBPFProgram(c *config.Config, offsets []manager.ConstantEditor, sockFD *
...
-		subprograms:     subprograms,
-		probesResolvers: subprogramProbesResolvers,
-		tailCallRouter:  tailCalls,
+		Manager:           errtelemetry.NewManager(mgr, bpfTelemetry),
+		cfg:               c,
+		offsets:           offsets,
+		subprograms:       subprograms,
+		probesResolvers:   subprogramProbesResolvers,
+		tailCallRouter:    tailCalls,
+		classificationMap: classificationMap,
```
Comment (nplanel) [inline pkg/network/protocols/http/ebpf_main.go:202]: We should have only one name, shall we rename `classificationMap` to `connectionMap` or rename `probes.ConnectionProtocolMap`

**Review by nplanel — APPROVED** [review]

**Review by richl-dd — APPROVED** [review]
Looks good for WKIT owned files.

**Pull request merged**